### ISSUE TITLE
conch tries to use Address objects as tuples

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5999]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)| @teratorn reported|
|-|-|
|Trac ID|trac#5999|
|Type|defect|
|Created|2012-09-18 00:34:53Z|
|Branch|https://github.com/twisted/twisted/tree/conch-channel-addresses-5999|

e.g. from a recent Exception in my code:
```
   File "/home/teratorn/code/ADE/lib/python2.6/site-packages/twisted/conch/ssh/channel.py", line 261, in getPeer
     return('SSH', )+self.conn.transport.getPeer()
 exceptions.TypeError: can only concatenate tuple (not "SSHTransportAddress") to tuple
```

I'm guessing that this regression was introduced when [#2997](https://github.com/twisted/twisted/issues/2997) was closed.
Perhaps it is doing this in other places too.

Attachments:

* [5999-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/db5/db5c69b3648ba7be49f589c259422579ee75684a/be30927d3c387da1041be6df9f88ad1b118d05c7.diff) (6253 bytes) - added by adiroiban on 2014-05-16 12:11:50Z - 
* [5339-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/db5/db5c69b3648ba7be49f589c259422579ee75684a/8fe6ff6fab9ccd84551f509cd909bcdeee4f7298.diff) (7350 bytes) - added by adiroiban on 2014-05-16 16:01:50Z - 
* [5999-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/db5/db5c69b3648ba7be49f589c259422579ee75684a/8cfec367ebefe1fdccbfcad1673b56e3d8529180.diff) (3853 bytes) - added by adiroiban on 2014-05-16 16:02:29Z - 
* [5999-3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/db5/db5c69b3648ba7be49f589c259422579ee75684a/b706a0518df1196120ee37f296377dbc65afc1ef.diff) (4391 bytes) - added by adiroiban on 2014-05-19 06:40:02Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5999 5999
type__defect defect
reporter__teratorn teratorn
priority__normal normal
milestone__ 
branch__branches_conch_channel_addresses_5999 branches/conch-channel-addresses-5999
branch_author__adiroiban__exarkun adiroiban, exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1347928493000000 1347928493000000
changetime__1448234214362545 1448234214362545
version__None None
owner__adiroiban adiroiban
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Both [#6305](https://github.com/twisted/twisted/issues/6305) and [#7261](https://github.com/twisted/twisted/issues/7261) were duplicates of this.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @adiroiban|
|-|-|

OK. I will try to fix this.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban removed owner|
|-|-|

Since getPeer was broken in channel, I have refactored it to return a simple
IAddress and not a tuple.

I have no idea why a tuple was needed, but a simple IAddress is consistent
with behavior of all other getPeer and getHost.

In case this still need to return a tuple, I think that  we should deprecate
this behavior.

I wrote a new test in channel for getHost and getPeer, one which uses a
real transport and not a mocked one.

In test_channel, MockTransport was only used for getPeer and getHost method
so I have removed it since it was replaced with a real transport.

I have removed the usage of SSHTransportAddress class since it seems over engineered
and not consistent with the other IAddress... I assume that SSHTranssport
is only done over TCP so IPv4 and IPv6 address should have the same methods.

Unfortunatly IAddress interface is just a marker interface and it does not
expose any methods.

Even though SSHTransport does not implements ITransport and is in fact
a protocol, I think that for
consistency is should try as much as possible to implement ITransport.

I don't know how to document simple/obvious methods which are already
documented in the interface.... adding a docstring,
return description and return type seems a waste of time and duplication.

Since the address module is no longer use maybe it should be deprecated.
Not sure if this should be done in this ticket, or open a new one.

Please check changes and let me know what needs to be changed or how I should
tackle this bug.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

> I have removed the usage of SSHTransportAddress class

This seems out of scope for this change.  I suggest removing this part of the patch so that the bug fix can land without being blocked by figuring out what is appropriate for `SSHTransportAddress`.

> since it seems over engineered and not consistent with the other IAddress...

I'm curious in what way you find it inconsistent.

> I assume that SSHTranssport is only done over TCP so IPv4 and IPv6 address should have the same methods.

I think that's a bad assumption. :)  And an SSH connection is *not* TCP/IPv4 or TCP/IPv6.  It doesn't seem like it is inherently a good idea to pretend that it is.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

Thanks for the initial review.

>> since it seems over engineered and not consistent with the other IAddress...
> I'm curious in what way you find it inconsistent.
>> I assume that SSHTranssport is only done over TCP so IPv4 and IPv6 address should have the same methods.
> I think that's a bad assumption. :) And an SSH connection is *not* TCP/IPv4 or TCP/IPv6. It doesn't seem like it is inherently a good idea to pretend that it is.

I was referring to SSHTransport, not the SSHConnection. I (still) guess that all SSH transports/connections will be done over TCP/IP. In this case, just returning an IPv4Address or IPv6Address makes much sense (for me) rather than returning an SSHTransportAddress which wraps an IPvXAddress.

When SSHTransportAddress is used, instead of just using calling transport.getPeer() (an in many other transports) I need to do transport.getPeer().address 

--------

Still, SSHChannel implements ITransport, so as described by the interface it should return an IAddress and not a tuple.


----------

I have attached a new patch which only touches the SSHChannel code.


Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

> When SSHTransportAddress is used, instead of just using calling transport.getPeer() (an in many other transports) I need to do transport.getPeer().address

Yes.  But there is a reason `getPeer` (and `getHost`) are only documented as returning an `IAddress` provider.  If we really only wanted to support IPv4 and IPv6 then we could simply document those methods as returning an `IPv4Address` or an `IPv6Address`.  But then we wouldn't be supporting SSH as a transport.

If you want your application to *really* work with arbitrary transports then you can't assume that `IAddress` means `IPv4Address` or `IPv6Address`.  But if your application only works with TCP over IPv4 or IPV6 then you wouldn't be trying to use Conch as your transport. :)

I suggest continuing this discussion elsewhere since it's not really related to this ticket.  I'd be happy to talk about the specifics of your application and how Twisted can be used or improved to support its requirements.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

Latest patch only touches the broken SSHChannel methods.

SSHChannel does implements ITransport so instead of returning a tuple, I think that it should return an IAddress, as described by the interface.

I have updated the initialization test of SSHChannel to check that it provides ITransport.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [43519]) Branching to 'conch-channel-addresses-5999'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [43520]) Apply 5999-3.diff 

Refs [#5999](https://github.com/twisted/twisted/issues/5999)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Thanks.  I checked the latest patch into a branch.  I also made a few minor changes:

  1. r43521 fixes the copyright header for the test module (pre-existing problem I noticed as I started this review)
  1. r43522 moves a new assertion added by the patch to a new test method (which was already asserting too many things)
  1. r43523 makes changes so that the tests don't need to do `channel.conn.transport = transport` and can instead re-use the existing terrible code that does that.
  1. r43524 makes minor whitespace changes to comply with the coding standard
  1. r43525 makes the two affected test method's docstrings follow the style guidelines and makes them more explicit about the desired behavior.
  1. r43526 reformats the news fragment to follow the style guidelines.
  1. r43527 is a trivial renaming of a variable used in one of the tests.

With this, I'm almost happy with the proposed change.  However, I suggest one additional change:

```
#!diff
Index: twisted/conch/test/test_channel.py
===================================================================
--- twisted/conch/test/test_channel.py  (revision 43527)
+++ twisted/conch/test/test_channel.py  (working copy)
@@ -287,7 +287,7 @@
         peer = IPv4Address('TCP', '192.168.0.1', 54321)
         self.connectTransport(peerAddress=peer)
 
-        self.assertEqual(peer, self.channel.getPeer())
+        self.assertEqual(SSHTransportAddress(peer), self.channel.getPeer())
 
 
     def test_getHost(self):
@@ -298,4 +298,4 @@
         host = IPv4Address('TCP', '127.0.0.1', 12345)
         self.connectTransport(hostAddress=host)
 
-        self.assertEqual(host, self.channel.getHost())
+        self.assertEqual(SSHTransportAddress(host), self.channel.getHost())
```

and the corresponding implementation change to make these modified tests pass, of course.

This is a slightly larger change than I'd prefer to make unilaterally as the reviewer (if someone else had proposed this I'd be happy to accept it though).  The implementation currently checked in to the branch hides the information that the transport is actually SSH over something else.

On the supporting side for this change, it's somewhat in line with what we've done for many of the recently introduced endpoint implementations (where we have lots of special address types for different kinds of endpoints).  This isn't quite a parallel since endpoints only use that address object as the argument to pass to `buildProtocol`.  Once the connection is set up, the transport is free to give back whatever values it wants from `getPeer` and `getHost`.  The point of this rich group of address types is to make it possible to tell exactly how things have been set up.  Conch provides an ITransport that pretty much works like a TCP connection but that doesn't mean it should be impossible for an application to easily discover it's not _just_ a TCP connection.

On the other hand, TLS transports don't return a special address type.  They just return whatever the underlying transport returns.  The way you figure out it's not just a TCP connection (or whatever) is by checking to see if the transport provides `ISSLTransport`.  This works... but it doesn't generalize.  Conch's SSH transport definitely shouldn't implement `ISSLTransport`...

Also in the negatives column is the fact that having your transport return an `SSHTransportAddress` object wrapped around an `IPv4Address` object _may_ just mean that suddenly applications that assumed TCP is the only thing are now broken (maybe they try accessing some IPv4Address attributes on the result of getPeer, for example).  Having a wild proliferation of types may not actually be a good way to express this additional information to applications (supposing we don't like the idea of breaking naive applications).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set owner to @adiroiban|
|-|-|

Not sure why I/we lost track of this. Assigning to myself so that I can continue working on it.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban removed owner|
|-|-|

Thanks for your changes!

I will move the debate over the mailinglist.

-----

While reading the code for getPeer and getHost I found an XXX comment so I have converted into a ticket and updated the comment.

-------

For twistedchecker errors there is this ticket https://github.com/twisted/twistedchecker/issues/44

--------

There is still something related to this change.

> The implementation currently checked in to the branch hides the information that the transport is actually SSH over something else.

Yes, but since you call getHost and getPeer on a SSHChannel I think that is obvious that it uses a SSHTransportBase.

Can we have a SSHChannel with a transport which is not based on SSHTransportBase ?

Instead of checking the address returned by the channel to see if you are using SSH you can just check the channel itself.

I am sending the branch to buildbots and for review.

Sorry for putting a fight on this but I don't see the point of using SSHTransportAddress.
Unfortunately the current documentation or ticket [#2997](https://github.com/twisted/twisted/issues/2997) does not explain why SSHTransportAddress is better than a normal IAddress.
I hope that feedback from the mailing list will help with the review of this ticket.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban commented|
|-|-|

I finally got why we need SSHTransportAddress.
I tried to update the docstring of SSHTransportAddress to explain why we need it at this stage so that others reading the code will not have a WTF moment.

I have pushed a change in which getHost and getPeer return as suggested by exarkun.

Please review.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

I ran the builders since I didn't see any current builds - they've picked up some twistedchecker errors.  No time for a full review right now, but maybe you can get the jump on these - https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/3013/steps/run-twistedchecker/logs/new%20twistedchecker%20errors

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_16"></a>|@adiroiban commented|
|-|-|

Fixed the doctrings. twistedchecker is green.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph removed owner|
|-|-|

Thanks for fixing this up, Adi.

I have two minor bits of feedback:

 1. The fact that `SSHChannel.getPeer` and `.getHost` return `SSHTransportAddress` is, I believe, a public implementation detail and should be documented in addition to their conformance with the (not terribly informative) `ITransport.getPeer` and `.getHost`, so I would put the `@rtype` back on those calls.
 2. `connectTransport` is a method on a `TestCase` subclass, but it actually doesn't depend on anything other than `self.channel.conn`.  Can you reduce its coupling and just make it a free function?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph set owner to @adiroiban|
|-|-|

Please just address these to your satisfaction and merge.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_19"></a>|@adiroiban commented|
|-|-|

Thanks for the review.

Somehow I lost track of this ticket.

1. Done. Added both return and rtype as twistedchecker was complaining that they are missing
2. Done.

Running tests after latest changes and will merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_20"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46323]) Merge conch-channel-addresses-5999: Fix return value of twisted.conch.ssh.channel.SSHChannel getPeer and getHost.

Author: adiroiban
Reviewer: exarkun, glyph
Fixes: [#5999](https://github.com/twisted/twisted/issues/5999)
